### PR TITLE
Make chart_test.go idempotent

### PR DESCRIPTION
**Pull request by RafalKorepta**
Due to the fact that in the charts/redpanda/Chart.yaml file the version needs to be bumped every time when Redpanda helm chart release happen, the charts/redpanda/Chart.lock in local environment will change over time. To make snapshot testing idempotent the `helm.sh/chart` label is overwritten during `helm templating`. During implementation of that feature it was discovered that merging of the commonLabels variable is wrongly used in `FullLabels` and the `statefulset-pod-labels` named template did not handle clash in user provided `commonLabels` for the following labels:
* app.kubernetes.io/name
* app.kubernetes.io/instance
* app.kubernetes.io/component

As the development of the test suite revealed that current merge function, that uses built-in maps package, does not work as the sprig equivalent, the helmette package tests where extended with the same test case as with sprig.

The gotohelm test suite also changed, so that any implementation of function that will be transpiled to helm template would not change the values in the dot context. During the course of merge function implementation the `values.commonLabels` was changed, so potentially other (next) usage of `commonLabels` would be malformed.

https://github.com/Masterminds/sprig/blob/581758eb7d96ae4d113649668fa96acc74d46e7f/docs/dicts.md?plain=1#L125-L182

https://github.com/Masterminds/sprig/blob/581758eb7d96ae4d113649668fa96acc74d46e7f/dict_test.go#L142-L197

### CODE REVIEW

**Review by chrisseto — APPROVED** [review]
All changes LGTM but I'm a bit confused about the underlying motivation. `Chart.Lock` is checked into source control, right? There shouldn't ever be any drift there of. We just need to update CI (I think an open PR from me does this) to actually run go test and then change the version bump action to update the golden files as well.

Or do you think "freezing" the version in the snapshot is preferable?

**Comment by chrisseto** [comment]
Ah, sorry I thought I added a commit message and marked this as non-draft but https://github.com/redpanda-data/helm-charts/pull/1112 should achieve a similar effect (Though it'll make automatic version bumps stop working for a bit).

The two PRs are compatible either way. I think I'm leaning towards your strategy as we'll reduce the total churn in the generated files. Can we change the version to something like `VERSIONPLACEHOLDER` in this PR so it's obvious that it's a placeholder value? If it needs to be semver maybe we set it to something like `9.9.9`?

**Comment by RafalKorepta** [comment]
Internally we agree that release process needs to change and Chart.yaml would have placeholders for `version` and `appVersion`. That way the labels would be stable for chart_test suite.

File: pkg/gotohelm/transpiler_test.go [hunk 1]
```diff
@@ -136,14 +144,16 @@ func TestTranspile(t *testing.T) {
 				spec.Values = append(spec.Values, map[string]any{})
 			}
 
-			for i, values := range spec.Values {
-				values := values
+			for i, originalValues := range spec.Values {
+				originalValues := originalValues
```
[the remark this answers is not in the document]
Comment (RafalKorepta) [reply pkg/gotohelm/transpiler_test.go:148]: Great point! I will adjust that logic.

File: pkg/gotohelm/transpiler_test.go [hunk 2]
```diff
@@ -197,6 +208,34 @@ func TestTranspile(t *testing.T) {
 	}
 }
 
+func changeAnyIntToFloat64(input map[string]any) map[string]any {
```
[the remark this answers is not in the document]
Comment (RafalKorepta) [reply pkg/gotohelm/transpiler_test.go:211]: Great point! I will adjust that logic.

File: charts/redpanda/helpers.go [hunk 3]
```diff
@@ -62,8 +62,34 @@ func FullLabels(dot *helmette.Dot) map[string]string {
...
 	}
 
-	// TODO Shouldn't our defaults take precedence over user provided labels?
-	return helmette.Merge(defaults, labels)
+	// As Merge function would not only return the dictionary, but also mutate its first argument
+	// the empty map is provided to not mutate user provided commonLabels
+	return helmette.Merge(map[string]string{}, labels, defaults)
+}
+
+func StatefulSetPodLabels(dot *helmette.Dot, ExistingStatefulSetLabels map[string]string) map[string]string {
```
[the remark this answers is not in the document]
Comment (RafalKorepta) [reply charts/redpanda/helpers.go:70]: You are right

File: charts/redpanda/helpers.go [hunk 4]
```diff
@@ -62,8 +62,34 @@ func FullLabels(dot *helmette.Dot) map[string]string {
...
+		return ExistingStatefulSetLabels
+	}
+	values := helmette.Unwrap[Values](dot.Values)
+
+	labels := map[string]string{}
+	if values.CommonLabels != nil {
+		labels = values.CommonLabels
+	}
+
+	component := fmt.Sprintf("%s-statefulset",
```
[the remark this answers is not in the document]
Comment (RafalKorepta) [reply charts/redpanda/helpers.go:81]: The 
https://github.com/redpanda-data/helm-charts/blob/e28537b17f3c66e731e9879ad4070767d1201d02/charts/redpanda/helpers.go#L174-L176

doesn't match with current implementation 
https://github.com/redpanda-data/helm-charts/blob/e28537b17f3c66e731e9879ad4070767d1201d02/charts/redpanda/templates/_statefulset.tpl#L30

51 characters != 63 characters

**Pull request closed without merging**